src/base/optimise/algodiff/owl_algodiff_generic.ml

```diff
@@ -82,41 +84,63 @@ module Make (A : Owl_types_ndarray_algodiff.Sig) = struct
   let jacobianTv' f x v =
     let x = make_reverse x (tag ()) in
     let y = f x in
+    if shape y <> shape v
+    then failwith "jacobianTv': vector not the same dimension as output";
     Reverse.reverse_reset y;
     Reverse.reverse_push v y;
-    primal y, x |> adjval |> primal
+    primal y, x |> adjval
 
 
   (* transposed jacobian vector product of f (vector -> vector) at x along v, backward ad *)
   let jacobianTv f x v = jacobianTv' f x v |> snd
 
   (* jacobian of f (vector -> vector) at x, both x and y are row vectors, also return the
      original value *)
-  let jacobian' f x =
-    let y = f x |> primal in
-    let m = col_num y in
-    let n = col_num x in
-    let z = A.empty [| m; n |] in
-    (match m > n with
-    | true ->
-      Array.init n (fun i ->
-          let v = A.zeros [| 1; n |] in
-          A.(set v [| 0; i |] (float_to_elt 1.));
-          jacobianv f x (Arr v))
-      |> Array.iteri (fun i v ->
-             match v with
-             | Arr v -> A.copy_col_to (A.transpose v) z i
-             | _ -> failwith "error: jacobian")
-    | false ->
-      Array.init m (fun i ->
-          let v = A.zeros [| 1; m |] in
-          A.(set v [| 0; i |] (float_to_elt 1.));
-          jacobianTv f x (Arr v))
-      |> Array.iteri (fun i v ->
-             match v with
-             | Arr v -> A.copy_row_to v z i
-             | _ -> failwith "error: jacobian"));
-    y, Arr z
+  let jacobian' =
+    let dim_typ x =
+      match primal' x with
+      | F _ -> `float
+      | Arr x ->
+        let s = A.shape x in
+        let d = Array.length s in
+        if d > 2
+        then `arr
+        else if s.(0) = 1
+        then `row s.(1)
+        else if s.(1) = 1
+        then `col s.(0)
+        else `mat
+      | _ -> assert false
```

Comment:
Can you raise a meaningful error here or this surely cannot happen?

Reply:
Yea I think with the definition of primal’, this surely cannot happen because it’s recursively extracting the primal... but yea we probably should design better helper functions to distinguish between these types